### ISSUE TITLE
JS hash fails to change when asset URL changes

### ISSUE THREAD

**Issue by jakearchibald** [rollup/rollup#3415]
- Rollup Version: 1.32.0
- Operating System (or Browser): OSX
- Node Version (if applicable): v12.16.1
- Link to reproduction: https://github.com/jakearchibald/rollup-hash-bug

In this demo, both `src1` and `src2` are built. The only difference between the two is the content of `some-asset.txt`.

### Expected Behavior

The hash of both the JS and asset is different between builds, since the JS imports the asset, the content of its file has changed.

### Actual Behavior

Only the hash of the txt file changes.

**Comment by lukastaegert** [comment]
Another important test case for a reimplemented hashing algorithm.

**Comment by jakearchibald** [comment]
This also impacts modules containing `import.meta.ROLLUP_FILE_URL_referenceId` that point to chunks.

With assets, you can kinda work around it with `augmentChunkHash`, using the buffer of the asset to augment the hash, but that isn't really possible with chunks, since they have dependencies of their own.

**Comment by sebastianseilund** [comment]
A workaround that seems to work for me is to add a hash of the asset's source as a JS comment together with `import.meta.ROLLUP_FILE_URL_referenceId`. So instead of the following from https://github.com/jakearchibald/rollup-hash-bug/blob/master/lib/asset-plugin.js :

```js
return `export default import.meta.ROLLUP_FILE_URL_${this.emitFile({
  type: 'asset',
  source: await fs.readFile(realId),
  name: basename(realId),
})}`;
```

Do this instead:

```js
import {createHash} from 'crypto'
// ...
const source = await fs.readFile(realId)
const referenceId = this.emitFile({
  type: 'asset',
  source,
  name: basename(realId),
})
const hash = createHash('sha256').update(source).digest('hex').slice(0, 8)
return `export default import.meta.ROLLUP_FILE_URL_${referenceId}; /* asset-hash:${hash} */`
```

Rollup will now consider the importing JS module's source as being different since the comment depends on the asset source hash, causing a different JS hash (which is what we want). The JS comment is stripped by minification later (if using).

**Comment by lukastaegert** [comment]
Fix at #4543

**Comment by lukastaegert** [comment]
Fix at #4543
